Source/WebCore/rendering/RenderListItem.cpp

```diff
@@ -231,25 +231,47 @@ void RenderListItem::updateValueNow() const
         }
     }
 
+    int defaultIncrement = (orderedList && orderedList->isReversed()) ? -1 : 1;
+    // FIXME: This should account for counter-reset as well.
+    auto valueForItem = [&](int previousValue, CounterDirectives& directives) {
+        if (directives.setValue)
+            return *directives.setValue;
+        return previousValue + directives.incrementValue.value_or(defaultIncrement);
+    };
+
     auto& startValue = startItem->m_value;
-    if (!startValue)
-        startValue = orderedList ? orderedList->start() : 1;
+    if (!startValue) {
+        auto directives = startItem->style().counterDirectives().map.get("list-item"_s);
+        startValue = valueForItem(orderedList ? orderedList->start() - defaultIncrement : 0, directives);
+    }
+
     int value = *startValue;
-    int increment = (orderedList && orderedList->isReversed()) ? -1 : 1;
 
     for (auto* item = startItem; item != this; ) {
         item = nextListItem(*list, *item);
-        item->m_value = (value += increment);
+        auto directives = item->style().counterDirectives().map.get("list-item"_s);
+        item->m_value = (value = valueForItem(value, directives));
     }
 }
 
 void RenderListItem::updateValue()
 {
-    if (!m_valueWasSetExplicitly) {
-        m_value = std::nullopt;
-        if (m_marker)
-            m_marker->setNeedsLayoutAndPrefWidthsRecalc();
-    }
+    m_value = std::nullopt;
+    if (m_marker)
+        m_marker->setNeedsLayoutAndPrefWidthsRecalc();
+}
+
+void RenderListItem::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle)
+{
+    RenderBlockFlow::styleDidChange(diff, oldStyle);
+
+    if (!oldStyle || (oldStyle->counterDirectives().map.isEmpty() && style().counterDirectives().map.isEmpty()))
```

Comment:
Why do we need this special case for when both maps are empty? Seems like the code below already will return efficiently in that case.